blockio: add blockio for convenient cgroup blkio control

askervin opened the pull request:
This is copied from the CRI Resource Manager with changes that drop dependencies to CRI Resource Manager configuration and cache.

This PR includes commits from PRs:
- PR #28 rdt: separate log wrapper to its own package
- PR #29 cgroups: add the cgroup filesystem interface

askervin commented:
@marquiz, finished sync with rdt annotation handling and earlier sync with SetConfig/SetConfigFromData/SetConfigFromFile.

marquiz commented:
Thanks @askervin! I think I need fresh eyes to review this so it'll be next week. Will get back to this on Monday.

marquiz commented on pkg/blockio/blockio.go:
```diff
@@ -0,0 +1,420 @@
+	classNames := make([]string, len(staticOciBlockIO))
+	i := 0
+	for name := range staticOciBlockIO {
+		i++
```
Shouldn't you do `i++` only after the assignment?

A way to avoid the counter is to create the slice with length 0
```
classNames := make([]string, 0, len(staticOciBlockIO))
```
and use append

askervin replied:
Well spotted, thanks! Gee... as a penance I added a unit test even for this simple function.

marquiz commented on pkg/blockio/blockio.go:
```diff
@@ -0,0 +1,420 @@
+	return oci, errors.ErrorOrNil()
+}
+
+// parseAndValidateInt64 parses quantities, like "64 M", and validates that they are in given range.
```
nit: I got confused by the name as it's handling positive integers. Rename to `parseAndValidateQuantity` or smth perhaps?

askervin replied:
Fixed.

marquiz reviewed (commented):
Thanks @askervin and apologies for the delayed review.

If I read this correctly, it is fairly tightly tied with OCI. But, can be used outside OCI as well (e.g. `SetCgroupClass()`). I just wonder could/should the code be split e.g. into two files so that the OCI-specific stuff would be in separate file? WDYT?

askervin commented on pkg/testutils/file.go:
```diff
@@ -0,0 +1,20 @@
+package testutils
```
Fixed

askervin commented on pkg/blockio/config.go:
```diff
@@ -0,0 +1,33 @@
+
+package blockio
+
+// Options captures our configurable parameters.
```
Fixed.

askervin commented on pkg/blockio/blockio.go:
```diff
@@ -0,0 +1,420 @@
+		// Handle all configurations as static for now. That
+		// is, the list of block devices matching Devices
+		// wildcards will not be updated without new
+		// configNotify(). class.DynamicDevices not supported
```
Fixed.

askervin commented:
@marquiz, thanks for the review!

In addition to the little changes, I did as you suggested, and split OCI-dependent code to `blockio/oci.go`.

I also removed all oci references from names in `cgroups/cgroupsblockio.go`, as they were misleading: not really OCI-related while looking very much the same, because both of these are in the level of cgroups abstraction.

Furthermore, I've added similar documentation to blockio as you have written to rdt. While testing the example code I noticed that blockio was logging irrelevant errors from reading Weights when I/O scheduler does not support them, even if user doesn't want to configure weights. Therefore I'm now skipping read errors in `cgroups/cgroupsblockio.ResetBlkioParameters`, and only report write errors - which should be sufficient.

I think this patch is now ready for re-review.

marquiz commented on pkg/blockio/blockio.go:
```diff
@@ -0,0 +1,456 @@
+	return SetConfig(config, force)
+}
+
+// SetConfig scans available block devices and applies new configuration
```
super nit: missing punctuation from the end of the sentence (just to be in line with the other "docstrings")

askervin replied:
Fixed!

marquiz commented on pkg/blockio/blockio.go:
```diff
@@ -0,0 +1,456 @@
+	return ios, nil
+}
+
+// deviceParametersToOci converts single blockio class parameters into OCI BlockIO structure.
```
s/deviceParametersToOci/devicesParametersToCgBlockIO/

askervin replied:
Fixed!

marquiz commented on pkg/blockio/blockio.go:
```diff
@@ -0,0 +1,456 @@
+// deviceParametersToOci converts single blockio class parameters into OCI BlockIO structure.
+func devicesParametersToCgBlockIO(dps []DevicesParameters, currentIOSchedulers map[string]string) (cgroups.BlockIOParameters, error) {
+	var errors *multierror.Error
+	oci := cgroups.NewBlockIOParameters()
```
Should this be renamed to something else than `oci`?

askervin replied:
Definitely, thanks! Fixed.

marquiz reviewed (commented):
Really great work @askervin! With the API documentation, tests and all 😍  

Basically ready to go. I had just a really small notes. One super-nit being about end of the sentence punctuation in the API documentation (add a period to all single-sentence docstrings) 😄

marquiz commented on pkg/blockio/blockio.go:
```diff
@@ -0,0 +1,456 @@
+	return nil
+}
+
+// GetClasses returns block I/O class names
```
nit: spotted one more missing punctuation from the public API documentation. This is the last nit like this, I promise 😊

askervin replied:
Thanks! Fixed. I should have went through all cases already in the previous patch. Now I've went through all the docstrings... :)

marquiz commented:
Looks really good! Let's merge this
